File: tree.py
class Node:
    def __init__(self, data):
        self.left = None
        self.right = None
        self.data = data

    def insert(self, data):
        tmp = Node(data)
        if self.data > data:
            if self.left != None:
                self.left.insert(data)
            else:
                self.left = tmp
        else:
            if self.right != None:
                self.right.insert(data)
            else:
                self.right = tmp
            
    def preorder(self):
        print (self.data, end=" ")
        if self.left != None: 
            self.left.preorder()
        if self.right != None:
            self.right.preorder()

    def inorder(self):
        if self.left != None:
            self.left.inorder()
        print (self.data, end=" ")
        if self.right != None:
            self.right.inorder()

    def postorder(self):
        if self.left != None:
            self.left.postorder()
        if self.right != None:
            self.right.postorder()
        print (self.data, end=" ")

    def find_val(self, val):
        if self.data > val:
            if self.left != None:
                self.left.find_val(val)
            else:
             print ("the value wasn't found") 
        elif self.data < val:
            if self.right != None:
                self.right.find_val(val)
            else:
                print ("the value wasn't found")
        else:
            print ("the value was found")
        

root = Node(10)
root.insert(5)
root.insert(20)
root.insert(2)
root.insert(7)
root.insert(18)
root.insert(17)

root.preorder()
print ("")
root.inorder()
print ("")
root.postorder()
print ("")

root.find_val(17)
root.find_val(40)

File: C++/tree.cpp
#include<iostream>
#include<stdlib.h>
#include<stdio.h>
#include<vector>
#include<stack>

using namespace std;

typedef struct Node {
    int value;
    Node *left;
    Node *right;

    Node(int value) {
        this->value = value;
        this->left =  NULL;
        this->right = NULL;
    }
} Node;

void insert(Node *root, Node *t) {

    if (root == NULL)
        root = t;

    if(t->value < root->value) {
        if(root->left != NULL) {
            insert(root->left, t);
        }
        else {
            root->left = t;
        }
    }
    else {
        if(root->right != NULL) {
            insert(root->right, t);
        }
        else {
            root->right = t;
        }
    }
}



void preorder(Node *root) {
    cout << root->value << " ";

    if (root->left != NULL) {
        preorder(root->left);
    }
    if(root->right != NULL) {
        preorder(root->right);
    }
}

void inorder_stack(Node *root) {
    stack<Node*> s;
    vector<int> result;
    
    Node *curr = root;

    while(curr != NULL || !s.empty()) {
        while(curr != NULL) {
            s.push(curr);
            curr = curr->left;
        }
        curr = s.top();
        s.pop();
        result.push_back(curr->value);
        curr = curr->right;

    }

    for(int i = 0; i < result.size(); i++) {
        cout << result[i] << " ";
    }

    cout << endl;
}

int main() {

    Node *root = new Node(7);
    Node *t;

    t = new Node(5);
    insert(root, t);

    t = new Node(10);
    insert(root, t);
   
    t = new Node(20);
    insert(root, t);

    

    cout << endl;
    return 0;
}
